Frameworks/include/CGImageInternal.h

```diff
@@ -96,31 +104,262 @@ typedef enum {
+    /*Alpha Only */
+    { GUID_WICPixelFormat8bppAlpha, { kCGColorSpaceModelRGB, (kCGImageAlphaOnly | kCGBitmapByteOrderDefault), 8, 32 } },
+    /* CYMK */
+    { GUID_WICPixelFormat32bppCMYK, { kCGColorSpaceModelCMYK, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 8, 32 } },
+    { GUID_WICPixelFormat64bppCMYK, { kCGColorSpaceModelCMYK, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 16, 64 } },
+    { GUID_WICPixelFormat40bppCMYKAlpha, { kCGColorSpaceModelCMYK, (kCGImageAlphaLast | kCGBitmapByteOrderDefault), 8, 40 } },
+    { GUID_WICPixelFormat80bppCMYKAlpha, { kCGColorSpaceModelCMYK, (kCGImageAlphaLast | kCGBitmapByteOrderDefault), 16, 64 } },
+    /*Monochrome*/
+    { GUID_WICPixelFormat4bppGray, { kCGColorSpaceModelMonochrome, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 4, 4 } },
+    { GUID_WICPixelFormat8bppGray, { kCGColorSpaceModelMonochrome, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 8, 8 } },
+    { GUID_WICPixelFormat16bppGray, { kCGColorSpaceModelMonochrome, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 16, 16 } },
+    { GUID_WICPixelFormat32bppGrayFloat, { kCGColorSpaceModelMonochrome, (kCGImageAlphaNone | kCGBitmapFloatComponents), 32, 32 } },
+    /*Indexed*/
+    { GUID_WICPixelFormat1bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 1 } },
+    { GUID_WICPixelFormat2bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 2 } },
+    { GUID_WICPixelFormat4bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 4 } },
+    { GUID_WICPixelFormat8bppIndexed, { kCGColorSpaceModelIndexed, (kCGImageAlphaNone | kCGBitmapByteOrderDefault), 1, 8 } }
+};
+
+struct __CGImageImpl {
+    Microsoft::WRL::ComPtr<IWICBitmapSource> bitmapImageSource;
+    bool isMask;
+    bool interpolate;
+    CGColorSpaceRef colorSpace;
+    CGColorRenderingIntent renderingIntent;
+
+    __CGImageImpl() {
+        isMask = false;
+        interpolate = false;
+        renderingIntent = kCGRenderingIntentDefault;
+    }
+
+    ~__CGImageImpl() {
+        if (colorSpace) {
+            CGColorSpaceRelease(colorSpace);
+        }
+    }
+
+    inline size_t Height() const {
+        size_t width, height;
+        RETURN_RESULT_IF_FAILED(bitmapImageSource->GetSize(&width, &height), 0);
+
+        return height;
+    }
+
+    inline size_t Width() const {
+        size_t width, height;
+        RETURN_RESULT_IF_FAILED(bitmapImageSource->GetSize(&width, &height), 0);
+
+        return width;
+    }
+
+    // TODO: Add cavet everywhere for how format changes
+    inline WICPixelFormatGUID PixelFormat() const {
+        WICPixelFormatGUID pixelFormat;
+        RETURN_RESULT_IF_FAILED(bitmapImageSource->GetPixelFormat(&pixelFormat), GUID_WICPixelFormatUndefined);
+        return pixelFormat;
+    }
+
+    inline const __CGImagePixelProperties* Properties() const {
```

Comment:
>Properties [](start = 43, length = 10)

store the value on first call, to be used later.
so we don't have to invoke the PixelFormat everytime.